File: clients/java/src/test/java/eu/autosignly/ContractTest.java
package eu.autosignly;

import com.fasterxml.jackson.annotation.JsonProperty;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.MethodSource;
import org.yaml.snakeyaml.Yaml;

import java.io.IOException;
import java.lang.reflect.RecordComponent;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.stream.Stream;

import static org.assertj.core.api.Assertions.assertThat;

/**
 * Contract test: every field name this client reads or sends must exist in the
 * published OpenAPI schema.
 *
 * <p>The field lists are not written by hand — they come from the record
 * components through reflection, so a model that gains or renames a field is
 * checked automatically, which a hand-kept list would never do.
 *
 * <p>Runs offline against {@code spec/autodocuments-v1.yaml}: no environment,
 * no network.
 */
class ContractTest {

    private static final Map<String, Object> SPEC = loadSpec();

    @SuppressWarnings("unchecked")
    private static Map<String, Object> loadSpec() {
        Path spec = Path.of("..", "..", "spec", "autodocuments-v1.yaml").toAbsolutePath().normalize();
        try {
            return (Map<String, Object>) new Yaml().load(Files.readString(spec));
        } catch (IOException e) {
            throw new IllegalStateException("Cannot read the API spec at " + spec, e);
        }
    }

    @SuppressWarnings("unchecked")
    private static Set<String> propertiesOf(String schema) {
        Map<String, Object> schemas =
                (Map<String, Object>) ((Map<String, Object>) SPEC.get("components")).get("schemas");
        Map<String, Object> found = (Map<String, Object>) schemas.get(schema);
        assertThat(found).as("schema %s is missing from the spec", schema).isNotNull();
        Map<String, Object> properties = (Map<String, Object>) found.get("properties");
        return properties == null ? Set.of() : new LinkedHashSet<>(properties.keySet());
    }

    /**
     * The wire names a record maps to, honouring an explicit {@code @JsonProperty}.
     *
     * <p>The annotation has to be looked up on the field and the accessor as well
     * as the component: {@code @JsonProperty} does not target record components,
     * so the compiler propagates it to the field, which is where Jackson reads it
     * from and where a naive reflection helper would miss it.
     */
    private static Set<String> wireNamesOf(Class<?> type) {
        Set<String> names = new LinkedHashSet<>();
        for (RecordComponent component : type.getRecordComponents()) {
            names.add(jsonNameOf(type, component));
        }
        return names;
    }

    private static String jsonNameOf(Class<?> type, RecordComponent component) {
        JsonProperty onComponent = component.getAnnotation(JsonProperty.class);
        if (onComponent != null) {
            return onComponent.value();
        }
        try {
            JsonProperty onField = type.getDeclaredField(component.getName()).getAnnotation(JsonProperty.class);
            if (onField != null) {
                return onField.value();
            }
        } catch (NoSuchFieldException ignored) {
            // A record always has the backing field; nothing to do if it is absent.
        }
        JsonProperty onAccessor = component.getAccessor().getAnnotation(JsonProperty.class);
        return onAccessor != null ? onAccessor.value() : component.getName();
    }

    static Stream<org.junit.jupiter.params.provider.Arguments> models() {
        return Stream.of(
                org.junit.jupiter.params.provider.Arguments.of(Models.Document.class, "DocumentInfoResponse"),
                org.junit.jupiter.params.provider.Arguments.of(Models.DocumentSummary.class, "DocumentListItemResponse"),
                org.junit.jupiter.params.provider.Arguments.of(Models.SignerDetails.class, "SignerResponse"),
                org.junit.jupiter.params.provider.Arguments.of(Models.SignerStatus.class, "SignerStatusResponse"),
                org.junit.jupiter.params.provider.Arguments.of(Models.Credentials.class, "CredentialsResponse"),
                org.junit.jupiter.params.provider.Arguments.of(Models.SigningRequestResult.class, "SendForSigningResponse"),
                org.junit.jupiter.params.provider.Arguments.of(Models.Tag.class, "TagResponse1"),
                org.junit.jupiter.params.provider.Arguments.of(Models.PageInfo.class, "PageInfo"),
                org.junit.jupiter.params.provider.Arguments.of(Models.Signer.class, "ExternalSignerRequest"));
    }

    @ParameterizedTest(name = "{0} matches {1}")
    @MethodSource("models")
    void modelOnlyUsesFieldsDeclaredInTheSchema(Class<?> model, String schema) {
        Set<String> declared = propertiesOf(schema);
        List<String> unknown = new ArrayList<>(wireNamesOf(model));
        unknown.removeAll(declared);

        assertThat(unknown)
                .as("%s uses fields the API does not know: %s", model.getSimpleName(), unknown)
                .isEmpty();
    }

    @Test
    void everyEndpointTheClientCallsExists() {
        @SuppressWarnings("unchecked")
        Set<String> paths = ((Map<String, Object>) SPEC.get("paths")).keySet();

        List<String> called = Arrays.asList(
                "/api/publics/v1/api-key",
                "/api/publics/v1/credentials",
                "/api/publics/v1/documents",
                "/api/publics/v1/documents/{documentId}",
                "/api/publics/v1/documents/signings",
                "/api/publics/v1/documents/{documentId}/send-for-signing",
                "/api/publics/v1/documents/{documentId}/tags",
                "/api/publics/v1/tags",
                "/api/publics/v1/tags/{tagId}");

        List<String> missing = new ArrayList<>(called);
        missing.removeAll(paths);

        assertThat(missing).as("endpoints gone from the API: %s", missing).isEmpty();
    }

    @Test
    void theSigningRequestSendsOnlyFieldsTheApiAccepts() {
        Set<String> declared = propertiesOf("UploadPdfAndSignRequest");

        // Built the same way the client builds it, so the assertion covers the
        // real request rather than a restatement of it.
        var body = AutosignlyClient.SigningOptions
                .of(List.of(Models.Signer.of("A", "B", "a@b.test", "PL")))
                .withSignature("SES", "SIGNATURES_CARD")
                .withVerificationMethod("SMS")
                .withInitiator("initiator@example.com", "pl")
                .toJson(new com.fasterxml.jackson.databind.ObjectMapper());
        body.put("documentName", "Umowa");

        List<String> unknown = new ArrayList<>();
        body.fieldNames().forEachRemaining(name -> {
            if (!declared.contains(name)) {
                unknown.add(name);
            }
        });

        assertThat(unknown).as("unknown request fields: %s", unknown).isEmpty();
        assertThat(propertiesOf("SigningInitiatorData")).contains("email", "locale");
    }
}

File: clients/node/src/contract.test.ts
/**
 * Contract test: every field name this client reads or sends must exist in the
 * published OpenAPI schema.
 *
 * The field lists are not written by hand — the parsers are called with a Proxy
 * that records which properties they touch. A parser that starts reading a
 * different key is therefore checked automatically, which is what a hand-kept
 * list would never do.
 *
 * This runs offline against `spec/autodocuments-v1.yaml`, so it needs no
 * environment and no network.
 */

import assert from "node:assert/strict";
import { readFileSync } from "node:fs";
import { join } from "node:path";
import { test } from "node:test";

import { parse as parseYaml } from "yaml";

import { AutosignlyClient } from "./client.js";
import {
  signerToPayload,
  toAttachment,
  toCredentials,
  toDocument,
  toDocumentSummary,
  toPage,
  toParty,
  toPartyAddress,
  partyToPayload,
  toSignerDetails,
  toSignerStatus,
  toSigningRequestResult,
  toTag,
} from "./models.js";

const SPEC_PATH = join(import.meta.dirname, "..", "..", "..", "spec", "autodocuments-v1.yaml");
const spec = parseYaml(readFileSync(SPEC_PATH, "utf8")) as {
  components: { schemas: Record<string, { properties?: Record<string, unknown> }> };
};

function propertiesOf(schema: string): Set<string> {
  const found = spec.components.schemas[schema];
  assert.ok(found, `schema ${schema} is missing from the spec`);
  return new Set(Object.keys(found.properties ?? {}));
}

/** Run a parser against a recording proxy and report which keys it read. */
function keysReadBy(parser: (payload: never) => unknown): string[] {
  const seen = new Set<string>();
  const recorder = new Proxy(
    {},
    {
      get(_target, property) {
        if (typeof property === "string") seen.add(property);
        return undefined;
      },
    },
  );
  parser(recorder as never);
  return [...seen];
}

const PARSERS: [string, string, (payload: never) => unknown][] = [
  ["toAttachment", "AttachmentResponse", toAttachment],
  ["toDocument", "DocumentInfoResponse", toDocument],
  ["toDocumentSummary", "DocumentListItemResponse", toDocumentSummary],
  ["toSignerDetails", "SignerResponse", toSignerDetails],
  ["toSignerStatus", "SignerStatusResponse", toSignerStatus],
  ["toCredentials", "CredentialsResponse", toCredentials],
  ["toSigningRequestResult", "SendForSigningResponse", toSigningRequestResult],
  ["toTag", "TagResponse", toTag],
  ["toParty", "Party", toParty],
  ["toPartyAddress", "PartyAddress", toPartyAddress],
];

for (const [name, schema, parser] of PARSERS) {
  test(`${name} only reads fields declared in ${schema}`, () => {
    const declared = propertiesOf(schema);
    const unknown = keysReadBy(parser).filter((key) => !declared.has(key));

    assert.deepEqual(unknown, [], `${name} reads fields the API does not send: ${unknown.join(", ")}`);
  });
}

test("toPage reads the envelope the API sends", () => {
  const envelope = propertiesOf("PageResponseDocumentListItemResponseV1");
  const unknown = keysReadBy((payload) => toPage(payload, () => null)).filter(
    (key) => !envelope.has(key),
  );

  assert.deepEqual(unknown, []);

  // The counters live one level down, in the nested page object.
  const info = propertiesOf("PageInfo");
  for (const counter of ["number", "size", "totalElements", "totalPages"]) {
    assert.ok(info.has(counter), `PageInfo lost ${counter}`);
  }
});

test("signerToPayload sends only fields the API accepts", () => {
  const declared = propertiesOf("ExternalSignerRequest");
  const sent = Object.keys(
    signerToPayload({
      firstName: "Anna",
      lastName: "Nowak",
      email: "anna@example.com",
      country: "PL",
      phoneNumber: "+48123456789",
      locale: "pl",
      order: 1,
      signatureType: "AES",
      signatureVerificationMethod: "SMS",
    }),
  );

  const unknown = sent.filter((key) => !declared.has(key));
  assert.deepEqual(unknown, [], `unknown signer fields: ${unknown.join(", ")}`);
});

test("partyToPayload sends only fields the API accepts", () => {
  const declared = propertiesOf("PartyRequest");
  const payload = partyToPayload({
    type: "COMPANY",
    name: "Acme Sp. z o.o.",
    firstname: "Jan",
    taxId: "5842831253",
    email: "kontakt@acme.pl",
    phone: "+48500100200",
    address: {
      street: "Marszalkowska",
      number: "12/34",
      postalCode: "00-001",
      city: "Warszawa",
      countryCode: "PL",
    },
  });

  const unknown = Object.keys(payload).filter((key) => !declared.has(key));
  assert.deepEqual(unknown, [], `unknown party fields: ${unknown.join(", ")}`);

  const address = propertiesOf("PartyAddress");
  const unknownAddress = Object.keys(payload.address as Record<string, unknown>).filter(
    (key) => !address.has(key),
  );
  assert.deepEqual(unknownAddress, []);
});

test("the party filters the client sends exist in the spec", () => {
  const parameters = (
    spec as unknown as {
      paths: Record<string, { get: { parameters: { name: string }[] } }>;
    }
  ).paths["/api/publics/v1/parties"].get.parameters;
  const declared = new Set(parameters.map((parameter) => parameter.name));

  for (const sent of ["name", "type", "page", "size"]) {
    assert.ok(declared.has(sent), `the API no longer accepts ${sent}`);
  }
});

test("uploadAndSign sends only fields the API accepts", async () => {
  let body: FormData | undefined;
  const client = new AutosignlyClient("k", "s", {
    baseUrl: "https://api.test/api",
    fetch: async (_url, init) => {
      body = init?.body as FormData;
      return new Response(JSON.stringify({ documentId: "d-1" }), {
        headers: { "content-type": "application/json" },
      });
    },
  });

  await client.uploadAndSign({
    pdf: new Uint8Array([37]),
    documentName: "Umowa",
    signatureType: "SES",
    signatureMode: "SIGNATURES_CARD",
    verificationMethod: "SMS",
    initiatorEmail: "initiator@example.com",
    signers: [{ firstName: "A", lastName: "B", email: "a@b.test", country: "PL" }],
  });

  const request = JSON.parse(await (body!.get("request") as Blob).text());
  const declared = propertiesOf("UploadPdfAndSignRequest");
  const unknown = Object.keys(request).filter((key) => !declared.has(key));
  assert.deepEqual(unknown, [], `unknown request fields: ${unknown.join(", ")}`);

  const initiator = propertiesOf("SigningInitiatorData");
  const unknownInitiator = Object.keys(request.signingInitiatorData).filter(
    (key) => !initiator.has(key),
  );
  assert.deepEqual(unknownInitiator, []);
});

test("the document filters the client sends exist in the spec", () => {
  const parameters = (
    spec as unknown as {
      paths: Record<string, { get: { parameters: { name: string }[] } }>;
    }
  ).paths["/api/publics/v1/documents"].get.parameters;
  const declared = new Set(parameters.map((parameter) => parameter.name));

  for (const sent of ["status", "tagId", "page", "size", "sort"]) {
    assert.ok(declared.has(sent), `the API no longer accepts ${sent}`);
  }
});

test("every endpoint the client calls exists in the spec", () => {
  const paths = new Set(Object.keys((spec as unknown as { paths: object }).paths));
  const called = [
    "/api/publics/v1/api-key",
    "/api/publics/v1/credentials",
    "/api/publics/v1/documents",
    "/api/publics/v1/documents/{documentId}",
    "/api/publics/v1/documents/{documentId}/attachments",
    "/api/publics/v1/documents/{documentId}/attachments/{attachmentId}",
    "/api/publics/v1/documents/signings",
    "/api/publics/v1/documents/{documentId}/signings",
    "/api/publics/v1/documents/{documentId}/tags",
    "/api/publics/v1/tags",
    "/api/publics/v1/tags/{tagId}",
    "/api/publics/v1/parties",
    "/api/publics/v1/parties/{partyId}",
  ];

  const missing = called.filter((path) => !paths.has(path));
  assert.deepEqual(missing, [], `endpoints gone from the API: ${missing.join(", ")}`);
});
